Initial draft of the full stack engineer job description (#114)
sourcegraph/careers: job-descriptions/fullstack-engineer.md

![logo](https://sourcegraph.com/.assets/img/sourcegraph-light-head-logo.svg)

> # Full Stack Engineer

Reviewer: ```suggestion
# Full Stack Engineer

```

Sourcegraph is seeking an engineer that thrives on owning big problems across domains and different levels of the stack — you grok how the ultimate measure of your work is how users experience the work you do, and that means you wear any hat necessary to that end, from product, to developer, and everything in between. You are a polyglot, quick learner and fearless into diving into unknown areas of the systems you work with. You believe that communicating clearly and empathetically and your relationships with others is critical to our success.

> ## Qualifications

Reviewer: ```suggestion

## Qualifications
```

> ## Qualifications

Author: That sounds like an engineer who is strong in one role and just tipping their toes in the water in the other (which is good) but to be considered a Full Stack engineer means that you are capable of developing full-fledged applications and have experience working any part of the stack. If they just started learning backend then they might be familiar with concepts but not possess functional knowledge - in that case they shouldn't be considered full stack.

We are looking for a full stack engineer who has strong fundamentals in good software development techniques, design patterns and best practices. In your career you have worked with and refactored existing code bases. You are both productive and pragmatic because you believe software is only useful if it is used. Collaborating with small high performing teams, you have built and deployed production-ready software that delivers value to customers.

> * Strong working knowledge with API design and architecture.

Reviewer: ```suggestion

* Strong working knowledge with API design and architecture.
```

* Skilled at building and testing end-to-end applications (e.g., unit testing, integration and e2e testing).
* Good knowledge of Git in particular and other version control systems in general.

> * Single page applications or progressive web apps and Web APIs.

Reviewer 2: The others are statements of wants, whereas this is a thing.  For example "Good knowledge of git..."; perhaps:

Experience developing single page applications or progressive web applications and APIs.

* Understanding of core web technologies (HTTP, HTML, CSS, Javascript)
* Strong experience with at least one web framework such as React, Angular, Polymer, Closure or Backbone.

> * Strong experience in at least one server side language such as Java, Python, PHP, Ruby.

Reviewer 2: Perhaps this could highlight the preference for Go and Typescript, but experience with Java, Python, etc. is also viable?

> * Strong experience in at least one server side language such as Java, Python, PHP, Ruby.

Author: We want a candidate that has strong knowledge of a popular server-side language, but concluded that most candidates won't have Go so we moved it to the 'nice to have' section and thought it was redundant to have it listed in two places.

 I'm ok with adding Go here as well.

* Experience in SQL, Postgres or other transactional, relational databases. Able to model and design schemas.
* Have worked with containers in production environments.

## Nice-to-haves

* Direct experience with any one of these technologies: ES6+, Go, GraphQL, Typescript, Docker, Kubernetes.
* Experience working on small high-performing teams, preferably tech startups.

> * Comfortable using a cloud computing platform such as AWS, GCP or Azure.

Reviewer 3: There are subjective terms used that we should replace with what they mean to us. What does skilled at building, strong experience, good knowledge, and comfortable mean to us? 

Examples within the the JD: 

> Have worked with containers in production environments.

Does a good job stating not just experience with containers, but having worked with them in production environments.

> Skilled at building and testing end-to-end applications (e.g., unit testing, integration and e2e testing).

Does a good job giving examples of what skilled means at the end. This can be rewritten as: 

> Experience building and testing end-to-end applications (e.g., unit testing, integration and e2e testing).

## Qualifications

We are looking for a full stack engineer who has strong fundamentals in good software development techniques, design patterns and best practices. In your career you have worked with and refactored existing code bases. You are both productive and pragmatic because you believe software is only useful if it is used. Collaborating with small high performing teams, you have built and deployed production-ready software that delivers value to customers.

* Strong working knowledge with API design and architecture.
* Experience building and testing end-to-end applications (e.g. unit testing, integration and e2e testing).
* Good knowledge of Git in particular and other version control systems in general.
* Experience building single page applications or progressive web apps and Web APIs.

> * Understanding of core web technologies (HTTP, HTML, CSS, Javascript)

Reviewer 4: ```suggestion
* Understanding of core web technologies (HTTP, HTML, CSS, JavaScript)
```

> * Strong experience with at least one web framework such as React, Angular, Polymer, Closure or Backbone.

Reviewer 4: ```suggestion
* Strong experience with at least one web framework such as React, Angular, Vue.js, Polymer or Closure.
```

Nit: trading in an old-fashioned framework for something a bit more current :)

* Strong experience in at least one server side language such as Java, Python, PHP, Ruby.
* Experience in SQL, Postgres or other transactional, relational databases. Able to model and design schemas.
* Have worked with containers in production environments.

## Nice-to-haves

> * Direct experience with any one of these technologies: ES6+, Go, GraphQL, Typescript, Docker, Kubernetes.

Reviewer 4: ```suggestion
* Direct experience with any one of these technologies: ES6+, Go, GraphQL, TypeScript, Docker, Kubernetes.
```

* Experience working on small high-performing teams, preferably tech startups.
* Comfortable using a cloud computing platform such as AWS, GCP or Azure.
* Published blog posts and/or tech talks about your work.
* Contributions to open source projects.

## Learn more about us

To create a product that serves the needs of all developers, we are building a diverse [all-remote team](https://about.sourcegraph.com/company/remote) that is [distributed across the world](https://about.sourcegraph.com/company/team). Sourcegraph is an equal opportunity workplace; we welcome people from all backgrounds and communities.

We provide [competitive compensation](https://about.sourcegraph.com/handbook/people-ops/compensation) and [practical benefits](https://about.sourcegraph.com/handbook/people-ops/benefits-and-perks) to keep you happy and healthy so that you can do your best work.

Learn more about what it is like to work at Sourcegraph by reading [our handbook](https://about.sourcegraph.com/handbook/).

## Interview process

> 1. You [apply here](TODO).

Reviewer 5: This needs a link

Reviewer: Added to TODO here: https://github.com/sourcegraph/interviews/pull/136#issue-468856668
